ci/pre-commit/example_links.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
+
+import os
+import json
+import git
+import re
+import argparse
+import zlib
+import base64
+import subprocess
+import sys
+
+def open_file(file_path):
+    with open(file_path, 'r') as file:
+        data = file.read()
+    return data
+
+def write_file(file_path, data):
+    with open(file_path, 'w') as file:
+        file.write(data)
+
+def create_compiler_explorer_url(file_path, COMPILER_ID, COMPILER_FLAGS):
+    source_code = open_file(file_path)
+    # Prepare the JSON payload
+    payload = {
+        "sessions": [
+            {
+                "id": 1,
+                "language": "cuda",
+                "source": source_code,
+                "compilers": [
+                    {
+                        "id": COMPILER_ID,
+                        "options": COMPILER_FLAGS,
+                        "libs": [
+                            {
+                                "id": "cccl",
+                                "version": "trunk"
+                            },
+                            {
+                                "id": "cuco",
+                                "version": "dev"
+                            }
+                        ]
+                    }
+                ]
+            }
+        ]
+    }
+
+    # Convert the payload to JSON
+    config_json = json.dumps(payload, separators=(',', ':'))
+
+    # Compress the JSON using zlib (deflate)
+    compressed = zlib.compress(config_json.encode('utf-8'))
+
+    # Base64 encode the compressed data
+    encoded = base64.urlsafe_b64encode(compressed).decode('utf-8')
+
+    # Replace problematic unicode characters
+    # As per the documentation, map characters in the range \u007F-\uFFFF
+    encoded = ''.join(
+        c if '\u0000' <= c <= '\u007F' else '\\u{:04x}'.format(ord(c))
+        for c in encoded
+    )
+
+    # Construct the final URL
+    url = f'https://godbolt.org/clientstate/{encoded}'
+
+    return url
+
+def update_readme_with_urls(readme_path, url_table, args):
+    # Open and read the README.md file
+    readme_content = open_file(readme_path)
+    lines = readme_content.splitlines()
+
+    updated_lines = []
+    mismatched_files = []
+    # Iterate through each line of the README
+    for line in lines:
+        updated_line = line
+        # Check each file in the url_table
+        for file_name, url in url_table.items():
```

Comment:
As discussed offline, we should throw/print an error in case an example source file has no matching line in the readme